### PR TITLE
Refactor `ContractCallServiceERCTokenTest`

### LINKED ISSUE #8456: Refactor `ContractCallServiceERCTokenTest `

### PR DESCRIPTION
**Pull request by nickeynikolovv**
**Description**:
With this PR the `ContractCallServiceERCTokenTest` integration test file is partially refactored using the web3j plugin and the big enum is now replaced with multiple junit tests so that it is easier to debug a particular scenario. This way also the number of persisted entities per test is reduced significantly and now we persist only the ones that are needed in the particular test. 

This PR refactors and extracts the read only erc contract functions in separate class, so it can be more readable. The modification functions will be extracted in another class with another PR

**Related issue(s)**:

Related and partially fixes #8456 

**Notes for reviewer**:

**Checklist**

- [ ] Documented (Code comments, README, etc.)
- [x] Tested (unit, integration, etc.)

### CODE REVIEW

File: hedera-mirror-web3/src/test/java/com/hedera/mirror/web3/service/ContractCallServiceERCTokenReadOnlyFunctionsTest.java [hunk 1]
```diff
@@ -0,0 +1,734 @@
...
+        final var tokenAddress = getAddressFromEntity(tokenEntity);
+        final var contract = testWeb3jService.deploy(ERCTestContract::deploy);
+        final var result = contract.call_getApprovedNonStatic(tokenAddress, BigInteger.valueOf(1)).send();
+        final var functionCall = contract.send_getApprovedNonStatic(tokenAddress, BigInteger.valueOf(1));
+        assertThat(result).isEqualTo((Address.ZERO).toHexString());
+        verifyEthCallAndEstimateGas(functionCall, contract);
+    }
+
+    @Test
+    void ethCallIsApporvedForAllStatic() throws Exception {
```
Comment (kselveliev) [inline hedera-mirror-web3/src/test/java/com/hedera/mirror/web3/service/ContractCallServiceERCTokenReadOnlyFunctionsTest.java:94]: Fix typo on method name

File: hedera-mirror-web3/src/test/java/com/hedera/mirror/web3/service/ContractCallServiceERCTokenTest.java [hunk 2]
```diff
@@ -59,21 +59,6 @@ private static Stream<Arguments> ercContractFunctionArgumentsProviderHistoricalR
     public static final String REDIRECT_SUFFIX = "Redirect";
     public static final String NON_STATIC_SUFFIX = "NonStatic";
 
-    @ParameterizedTest
-    @MethodSource("ercContractFunctionArgumentsProvider")
-    void ercReadOnlyPrecompileOperationsTest(final ErcContractReadOnlyFunctions ercFunction, final boolean isStatic) {
```
Comment (kselveliev) [inline hedera-mirror-web3/src/test/java/com/hedera/mirror/web3/service/ContractCallServiceERCTokenTest.java:64]: Here we delete the method that tests eth_call.
But in our newly created suite we test for eth call and estimate gas response.
From what I see there is another method with the same parameter source that tests for estimate gas called 
`supportedErcReadOnlyPrecompileOperationsTest` i believe we could also delete that if im not mistaken

Comment (nickeynikolovv) [reply hedera-mirror-web3/src/test/java/com/hedera/mirror/web3/service/ContractCallServiceERCTokenTest.java:64]: Correct, removed

File: hedera-mirror-web3/src/test/java/com/hedera/mirror/web3/service/ContractCallServiceERCTokenReadOnlyFunctionsTest.java (hunk 1 above)
Comment (nickeynikolovv) [reply hedera-mirror-web3/src/test/java/com/hedera/mirror/web3/service/ContractCallServiceERCTokenReadOnlyFunctionsTest.java]: fixed

File: hedera-mirror-web3/src/test/java/com/hedera/mirror/web3/service/ContractCallServiceERCTokenReadOnlyFunctionsTest.java [hunk 3]
```diff
@@ -0,0 +1,734 @@
...
+
+class ContractCallServiceERCTokenReadOnlyFunctionsTest extends AbstractContractCallServiceTest {
+
+    public ContractCallServiceERCTokenReadOnlyFunctionsTest(TestWeb3jService testWeb3jService) {
+        super(testWeb3jService);
+    }
+
+    @Test
+    void ethCallGetApprovedEmptySpenderStatic() throws Exception {
+      final var treasuryEntityId = accountPersist();
```
[the remark this answers is not in the document]
Comment (nickeynikolovv) [reply hedera-mirror-web3/src/test/java/com/hedera/mirror/web3/service/ContractCallServiceERCTokenReadOnlyFunctionsTest.java]: fixed

File: hedera-mirror-web3/src/test/java/com/hedera/mirror/web3/service/ContractCallServiceERCTokenTest.java [hunk 4]
```diff
@@ -306,20 +269,20 @@ public String getName(final boolean isStatic) {
...
-        APPROVE_NFT("approve", new Object[] {NFT_ADDRESS, SPENDER_ADDRESS, 1L}),
-        APPROVE_WITH_ALIAS("approve", new Object[] {FUNGIBLE_TOKEN_ADDRESS, SENDER_ALIAS, 2L}),
-        TRANSFER("transfer", new Object[] {TREASURY_TOKEN_ADDRESS, SPENDER_ALIAS, 2L}),
-        TRANSFER_FROM("transferFrom", new Object[] {TREASURY_TOKEN_ADDRESS, SENDER_ALIAS, SPENDER_ALIAS, 2L}),
+        APPROVE("approve", new Object[]{FUNGIBLE_TOKEN_ADDRESS, SPENDER_ALIAS, 2L}),
+        DELETE_ALLOWANCE_NFT("approve", new Object[]{NFT_ADDRESS, Address.ZERO, 1L}),
+        APPROVE_NFT("approve", new Object[]{NFT_ADDRESS, SPENDER_ADDRESS, 1L}),
+        APPROVE_WITH_ALIAS("approve", new Object[]{FUNGIBLE_TOKEN_ADDRESS, SENDER_ALIAS, 2L}),
+        TRANSFER("transfer", new Object[]{TREASURY_TOKEN_ADDRESS, SPENDER_ALIAS, 2L}),
+        TRANSFER_FROM("transferFrom", new Object[]{TREASURY_TOKEN_ADDRESS, SENDER_ALIAS, SPENDER_ALIAS, 2L}),
```
Comment (steven-sheehy) [inline hedera-mirror-web3/src/test/java/com/hedera/mirror/web3/service/ContractCallServiceERCTokenTest.java:277]: It looks like spotless is not running. Please run `./gradlew spotlessApply` and ensure going forward the git hook executes even if you have to change from committing via a UI to `git commit`.

File: hedera-mirror-web3/src/test/java/com/hedera/mirror/web3/service/ContractCallServiceERCTokenReadOnlyFunctionsTest.java [hunk 5]
```diff
@@ -0,0 +1,756 @@
+/*
+ * Copyright (C) 2023-2024 Hedera Hashgraph, LLC
```
Comment (steven-sheehy) [inline hedera-mirror-web3/src/test/java/com/hedera/mirror/web3/service/ContractCallServiceERCTokenReadOnlyFunctionsTest.java:2]:
```suggestion
 * Copyright (C)2024 Hedera Hashgraph, LLC
```

File: hedera-mirror-web3/src/test/java/com/hedera/mirror/web3/service/ContractCallServiceERCTokenReadOnlyFunctionsTest.java [hunk 6]
```diff
@@ -0,0 +1,756 @@
...
+    }
+
+    private EntityId accountPersist() {
+        return domainBuilder
+                .entity()
+                .customize(e -> e.evmAddress(null)
+                        .balance(12L))
+                .persist()
+                .toEntityId()
+                ;
```
Comment (steven-sheehy) [inline hedera-mirror-web3/src/test/java/com/hedera/mirror/web3/service/ContractCallServiceERCTokenReadOnlyFunctionsTest.java:680]:
```suggestion
                .toEntityId();
```

File: hedera-mirror-web3/src/test/java/com/hedera/mirror/web3/service/ContractCallServiceERCTokenReadOnlyFunctionsTest.java [hunk 7]
```diff
@@ -0,0 +1,756 @@
...
+        final var tokenEntity = persistTokenEntity();
+        domainBuilder
+                .token()
+                .customize(t -> t.tokenId(tokenEntity.getId())
+                        .type(TokenTypeEnum.NON_FUNGIBLE_UNIQUE))
+                .persist();
+        domainBuilder
+                .nft()
+                .customize(n -> n.tokenId(tokenEntity.getId()).serialNumber(1L))
+                .persist();
```
Comment (steven-sheehy) [inline hedera-mirror-web3/src/test/java/com/hedera/mirror/web3/service/ContractCallServiceERCTokenReadOnlyFunctionsTest.java:566]: This sort of setup is in a number of tests. Might be useful to extract it to a helper method.

**Review by steven-sheehy — CHANGES_REQUESTED** [review]

File: hedera-mirror-web3/src/test/java/com/hedera/mirror/web3/service/ContractCallServiceERCTokenReadOnlyFunctionsTest.java [hunk 8]
```diff
@@ -0,0 +1,756 @@
...
+import com.hedera.mirror.web3.web3j.generated.ERCTestContract;
+import com.hedera.services.store.models.Id;
+import com.hedera.services.utils.EntityIdUtils;
+import java.math.BigInteger;
+import org.hyperledger.besu.datatypes.Address;
+import org.junit.jupiter.api.Test;
+
+class ContractCallServiceERCTokenReadOnlyFunctionsTest extends AbstractContractCallServiceTest {
+
+    public ContractCallServiceERCTokenReadOnlyFunctionsTest(TestWeb3jService testWeb3jService) {
```
Comment (bilyana-gospodinova) [inline hedera-mirror-web3/src/test/java/com/hedera/mirror/web3/service/ContractCallServiceERCTokenReadOnlyFunctionsTest.java:43]: The modifier needs to be private as we don't need instances of this class.

Comment (nickeynikolovv) [reply hedera-mirror-web3/src/test/java/com/hedera/mirror/web3/service/ContractCallServiceERCTokenReadOnlyFunctionsTest.java]: fixed

File: hedera-mirror-web3/src/test/java/com/hedera/mirror/web3/service/ContractCallServiceERCTokenReadOnlyFunctionsTest.java (hunk 7 above)
Comment (nickeynikolovv) [reply hedera-mirror-web3/src/test/java/com/hedera/mirror/web3/service/ContractCallServiceERCTokenReadOnlyFunctionsTest.java]: fixed - extracted in a separate method

File: hedera-mirror-web3/src/test/java/com/hedera/mirror/web3/service/ContractCallServiceERCTokenReadOnlyFunctionsTest.java (hunk 6 above)
Comment (nickeynikolovv) [reply hedera-mirror-web3/src/test/java/com/hedera/mirror/web3/service/ContractCallServiceERCTokenReadOnlyFunctionsTest.java]: fixed

File: hedera-mirror-web3/src/test/java/com/hedera/mirror/web3/service/ContractCallServiceERCTokenReadOnlyFunctionsTest.java (hunk 5 above)
Comment (nickeynikolovv) [reply hedera-mirror-web3/src/test/java/com/hedera/mirror/web3/service/ContractCallServiceERCTokenReadOnlyFunctionsTest.java]: fixed

File: hedera-mirror-web3/src/test/java/com/hedera/mirror/web3/service/ContractCallServiceERCTokenTest.java (hunk 4 above)
Comment (nickeynikolovv) [reply hedera-mirror-web3/src/test/java/com/hedera/mirror/web3/service/ContractCallServiceERCTokenTest.java]: done

**Review by steven-sheehy — COMMENTED** [review]
LG, just one code smell and need to fix merge conflicts

**Review by bilyana-gospodinova — APPROVED** [review]
LGTM

**Review by kselveliev — APPROVED** [review]

**Review by steven-sheehy — APPROVED** [review]

**Pull request merged**